File: go/space-age/space_age.go
// Package space calculates how old someone would be on a given planet by their
// age in earth seconds
package space

// Planet is a type of string
type Planet string

// Age converts earth seconds to given planet's years
func Age(seconds float64, planet Planet) float64 {
	var base float64 = 31557600

	age := func(year float64) float64 {
		return seconds / (base * year)
	}

	switch planet {
	case "Earth":
		return age(1)
	case "Mercury":
		return age(0.2408467)
	case "Venus":
		return age(0.61519726)
	case "Mars":
		return age(1.8808158)
	case "Jupiter":
		return age(11.862615)
	case "Saturn":
		return age(29.447498)
	case "Uranus":
		return age(84.016846)
	case "Neptune":
		return age(164.79132)
	default:
		return age(1)
	}
}

File: ruby/space-age/space_age.rb
# SpaceAge calculates how old someone would be on different planets from seconds
class SpaceAge
  EARTH_ORBITAL = 315_576_00.0 # seconds

  PLANETS = {
    earth: 1.0,
    mercury: 0.2408467,
    venus: 0.61519726,
    mars: 1.8808158,
    jupiter: 11.862615,
    saturn: 29.447498,
    uranus: 84.016846,
    neptune: 164.79132
  }.freeze

  PLANETS.each do |planet, earth_years|
    method_name = "on_#{planet}"
    define_method(method_name) do
      calculate_age(earth_years)
    end
  end

  attr_reader :seconds

  def initialize(seconds)
    @seconds = seconds
  end

  def calculate_age(earth_years)
    seconds / (EARTH_ORBITAL * earth_years)
  end
end
